docs/docs/guides/tracking/costs.md

```diff
@@ -0,0 +1,163 @@
+The two required fields are `prompt_token_cost` and `completion_token_cost`.
+You can also set `effective_date` to a datetime, to make the cost effective at a specific date, this defaults to the current date.
+
+```python
+import weave
+from datetime import datetime
+
+client = weave.init("my_custom_cost_model")
+
+client.add_costs({
+    "your_model_name": {
+        "prompt_token_cost": 0.1,
+        "completion_token_cost": 0.2,
+    }
+})
+
+client.add_costs({
+    "your_model_name": {
+        "prompt_token_cost": 10,
+        "completion_token_cost": 20,
+        # If for example I want to raise the price of the model after a certain date
+        "effective_date": datetime(2025, 4, 22),
+    },
+    "my_special_model_1": {
+        "prompt_token_cost": 0.1,
+        "completion_token_cost": 0.2,
+        "effective_date": datetime(1972, 5, 11),
+    }
+})
+```
+
+## Querying for costs
+
+You can query for costs by using the [`query_costs`](/reference/python-sdk/weave/trace/weave.trace.weave_client#method-query_costs) method.
+There are a few ways to query for costs, you can pass in a singular cost id, or a list of LLM model names.
+
+```python
+import weave
+
+client = weave.init("my_custom_cost_model")
+
+costs = client.query_costs(llm_ids=["your_model_name"])
+
+cost = client.query_costs(costs[0].id)
+```
+
+## Purging a custom cost
+
+You can purge a custom cost by using the `purge_costs` method. You pass in a list of cost ids, and the costs with those ids are purged.
+
+```python
+import weave
+
+client = weave.init("my_custom_cost_model")
+
+costs = client.query_costs(llm_ids=["your_model_name"])
+client.purge_costs([cost.id for cost in costs])
+```
+
+## Setting up a custom cost model
+
+Weave calculates costs based on the number of tokens used and the model used.
+Weave grabs this usage and model from the output and associates them with the call.
+
+Let's set up a simple custom model, that calculates its own token usage, and stores that in weave.
+
+```python
+from weave import Model
+import weave
+from weave.trace_server.trace_server_interface import CallsFilter
+
+class YourModel(Model):
+    attribute1: str
+    attribute2: int
+
+    def simple_token_count(self, text: str) -> int:
+        return len(text) // 3
+
+    # This is a custom op that we are defining
+    # It takes in a string, and outputs a dict with the usage counts, model name, and the output
+    @weave.op()
+    def custom_model_generate(self, input_data: str) -> dict:
+        # Model logic goes here
+        # Here is where you would have a custom generate function
+        prediction = self.attribute1 + ' ' + input_data
+
+        # Usage counts
+        prompt_tokens = self.simple_token_count(input_data)
+        completion_tokens = self.simple_token_count(prediction)
+
+        return {"usage": {'input_tokens': prompt_tokens, 'output_tokens': completion_tokens}, "model": 'your_model_name', 'output': prediction,}
```

Comment:
I would explain this data structure - comment that it conforms to openais output